### PR TITLE
Surface graph pre-encoder (GINO-style) for τ_y/τ_z structural gaps

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The dominant open gaps are τ_y (3.41×) and τ_z (3.65×) vs AB-UPT targets. These are geometric quantities — wall shear stress components depend critically on the local surface orientation, curvature, and how flow detaches/reattaches along the vehicle body. The current Transolver backbone processes surface points as an unordered token set without explicit awareness of the geometric manifold structure.

**We hypothesize that a Geometry-Informed Neural Operator (GINO) approach will outperform the current backbone by:**
1. Lifting input features into a learned latent function on a **graph** over the surface point cloud (using radius-graph edges), giving the model explicit local geometric structure not available to Transolver's global slice attention
2. Applying **message-passing layers on the surface graph** as a pre-encoder before the Transolver backbone (hybrid approach) — this is lower-risk than a full backbone replacement and preserves all the compounding wins (learnable PE, β-NLL, Lion)
3. The graph pre-encoder aggregates **local geometry context** (neighbour positions, edge vectors, local surface normals) into per-point features, which then feed into the existing Transolver as richer tokens

This is inspired by GINO (Li et al., 2023: https://arxiv.org/abs/2309.00583) and GNO (Li et al., 2020). The key insight: for τ_y/τ_z prediction, what matters most is the **local geometric neighborhood** (incident angle, curvature, adjacent surface elements) which a global attention model can only learn implicitly. A graph message-passing pre-encoder provides this explicitly.

**Implementation strategy (lower risk than full replacement):**
- Keep the existing Grouped Transolver backbone completely intact
- Add an optional `SurfaceGraphEncoder` module that runs **before** the surface tokens enter the Transolver
- Use `torch_geometric` (or pure PyTorch scatter ops to avoid dependency) for radius-graph construction and message passing

## Instructions

Add a `SurfaceGraphEncoder` class to `train.py` and a `--surface-graph-encoder` flag to enable it. The encoder wraps the surface input only; volume stream is unchanged.

**Radius graph construction:**
- Build a radius graph over surface points: for each point, find all neighbours within radius `r` (suggest `r=0.05` in normalized coordinates — tune if connectivity is too sparse/dense)
- Use k-NN as fallback if radius graph is too sparse: `k=16` nearest neighbours
- Edge features: relative position vector (Δx, Δy, Δz) + L2 distance (4D edge feature)
- To avoid OOM on 65k-point clouds: **chunk the graph construction** over mini-batches of 4096 points at a time, or use approximate kNN (faiss-style) if available. Alternatively, build the graph on a **subsample** of 8192 surface points and use scatter-interpolate to lift back to 65k.

**Recommended approach (memory-efficient):**
1. Subsample surface to 8192 points (random each step)
2. Build kNN graph (k=16) on the 8192-point subset: cost = 8192² manageable
3. Run 2–3 message-passing layers (EdgeConv or simple sum-aggregation MLP)
4. Scatter/interpolate the graph features back to all 65k surface points using nearest-neighbour lookup
5. Concatenate with original surface features → feed enriched tokens into Transolver

**EdgeConv message passing (simple, effective):**
```python
# For each edge (i, j): m_ij = MLP(cat(h_i, h_j - h_i, e_ij))
# h_i_new = mean(m_ij for j in neighbours(i))
```

**CLI flags to add:**
```
--surface-graph-encoder         # enable the graph pre-encoder (default: off)
--graph-encoder-layers 2        # number of MP layers (default: 2)
--graph-encoder-hidden 256      # hidden dim for graph MLP (default: 256)
--graph-knn 16                  # k for kNN graph (default: 16)
--graph-subsample 8192          # subsample size for graph construction (default: 8192)
```

**Training command (Arm A — baseline config + graph encoder):**
```bash
torchrun --standalone --nproc_per_node=4 train.py \
  --agent edward \
  --surface-graph-encoder \
  --graph-encoder-layers 2 \
  --graph-encoder-hidden 256 \
  --graph-knn 16 \
  --graph-subsample 8192 \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --clip-grad-norm 0.5 \
  --no-compile-model \
  --learnable-pe \
  --beta-nll-beta 0.5 \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --model-slices 128 \
  --ema-decay 0.999 \
  --lr-warmup-epochs 1 \
  --wandb_group edward-gino-graph-encoder
```

Note: `--beta-nll-beta 0.5` is already on yi (your own PR #583) — stack it with the graph encoder.

**Arm B (if Arm A shows promise, i.e. val_abupt < 9.5% after 2 epochs):**
Increase `--graph-encoder-layers 3 --graph-encoder-hidden 512` and `--graph-knn 32`.

**OOM mitigation:** If 65k points + graph encoder OOMs, reduce `--graph-subsample 4096` or `--train-surface-points 32768`. The graph runs on the subsample, so reducing subsample size is much cheaper than reducing the full point cloud.

**Important:** Use `--no-compile-model` since `--learnable-pe` requires it (torch.compile inductor broadcast bug). The graph encoder itself should not cause additional compile issues.

## Baseline to beat

Current yi SOTA: **val_abupt = 8.861%** (PR #583 edward, β-NLL beta=0.5, W&B run `5xovw2si`)

| Metric | Current yi best (val, PR #583) | AB-UPT target | Gap |
|---|---:|---:|---:|
| abupt_axis_mean_rel_l2_pct | 8.861% | — | — |
| surface_pressure_rel_l2_pct | 5.457% | 3.82 | 1.43× |
| wall_shear_rel_l2_pct | 10.022% | 7.29 | 1.37× |
| volume_pressure_rel_l2_pct | 4.951% | 6.08 | ✓ |
| wall_shear_y_rel_l2_pct | 12.447% | 3.65 | **3.41×** |
| wall_shear_z_rel_l2_pct | 13.235% | 3.63 | **3.65×** |

Reproduce current SOTA:
```bash
torchrun --standalone --nproc_per_node=4 train.py \
  --agent edward --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --clip-grad-norm 0.5 --no-compile-model --learnable-pe \
  --beta-nll-beta 0.5 --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --model-slices 128 --ema-decay 0.999 --lr-warmup-epochs 1
```

## Terminal result format

When done (either Arm A alone, or Arm A + Arm B), post a comment with a single-line marker:
```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<arm-a-run-id>"],"primary_metric":{"name":"val/primary/abupt_axis_mean_rel_l2_pct","value":<number>},"test_metric":{"name":"test/primary/abupt_axis_mean_rel_l2_pct","value":<number>}}
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: edward — spec looks good, implementation is well-defined. Start with Arm A now.

## Implementation checklist

1. **EdgeConv block** — implement `EdgeConvBlock(in_dim, out_dim, edge_feat_dim=4)`:
   - Edge features: `[Δx, Δy, Δz, ||Δx||₂]` (4D) between point i and neighbor j
   - Message: `m_ij = MLP(cat(h_i, h_j − h_i, e_ij))`, `h_i_new = mean(m_ij over k neighbors)`
   - MLP: `Linear(in_dim + in_dim + 4 → hidden) → ReLU → Linear(hidden → out_dim)`

2. **Surface graph pre-encoder** — new class `SurfaceGraphEncoder`:
   - Input: surface points with features `(B, N_surf, C_in)`
   - Subsample: `fps_or_random_subsample(N_surf, 8192)` for speed (random is fine for now)
   - Build kNN graph: `k=16` neighbors using L2 distance in xyz space
   - Apply `graph_encoder_layers` (default=2) EdgeConv blocks: hidden=`graph_encoder_hidden` (default=256)
   - Scatter back: for each of the 65k surface points, scatter the feature of its nearest subsampled point using `torch_scatter` or a simple gather with pre-computed indices
   - Output: concat original surface features + graph-encoded features

3. **Integration into main model** — `--surface-graph-encoder` flag gates the entire module. When enabled, surface point features are expanded before being passed into Transolver.

## Training command — Arm A

```bash
python train.py \
  --model transolver \
  --num-heads 8 --num-slices 32 --depth 4 --hidden-dim 512 \
  --learnable-pe --no-compile-model \
  --beta-nll-beta 0.5 \
  --surface-graph-encoder \
  --graph-encoder-layers 2 \
  --graph-encoder-hidden 256 \
  --graph-knn 16 \
  --graph-subsample 8192 \
  --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --batch-size 4 \
  --wandb-group edward-surface-graph-encoder
```

## Gates

- **Arm A EP1 val_abupt < 20%**: proceed (viability check — GINO encoder is not crashing training)
- **Arm A EP2 val_abupt < 10.5%**: promising — launch Arm B (layers=3, hidden=512, knn=32)
- **Arm A EP2 val_abupt < 9.032%**: merge-ready — post terminal SENPAI-RESULT
- **Arm A EP2 val_abupt ≥ 10.5%**: report per-axis breakdown (especially τ_y, τ_z); I'll decide if Arm B is warranted

## Critical diagnostic to report
At each epoch checkpoint report: val_abupt, τ_y, τ_z, cp, vol_p AND the fraction of 65k surface points that had a valid subsampled neighbor within k=16 (sanity check on the scatter coverage — should be >99%).

Current yi baseline to beat: **9.032%** (PR #517, W&B `brat65z4`).
Best in-flight: thorfinn #590 Arm B' EP2 = 9.672%, tracking toward merge.

Good luck.

**Comment by morganmcg1** [comment]
ADVISOR: Closing this PR — EP1 val_abupt = 28.751% on run `72bxqlnu` (step 5,449) catastrophically fails the viability gate I set (< 20%). The GINO surface graph pre-encoder (EdgeConv k=16, FPS→8,192, scatter→65k) is not helping convergence at this parameter budget.

Root cause analysis: the pre-encoder adds ~8M parameters and a full kNN graph build on 65k surface points. At 1 epoch of training, the EdgeConv weights are essentially random — there is no inductive bias benefit, only noise injection into the slice tokens. The scatter aggregation from 8,192 FPS nodes back to 65k surface points compounds the interpolation error.

Lesson learned: surface graph pre-encoders need at least 2–3 epochs to learn meaningful local geometry features before they can help downstream slice-token attention. At our 1-epoch EP1 viability gate this approach will always fail unless the encoder is pre-trained or initialized from a geometry-aware checkpoint.

Thank you for the careful implementation. Closing now to free the GPU for higher-value experiments.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #617 because Closing — EP1 val_abupt = 28.751% (run 72bxqlnu) fails viability gate (< 20%). GINO surface graph pre-encoder adds noise at 1-epoch budget; EdgeConv weights are untrained at EP1. Lesson: surface GNN pre-encoders require multi-epoch warm-up to be beneficial. Advisory comment posted above..
